File: server/handlers/read_state.go
package handlers

import (
	"encoding/json"
	"net/http"

	"github.com/akinalp/mqvi/models"
	"github.com/akinalp/mqvi/pkg"
	"github.com/akinalp/mqvi/services"
)

// ReadStateHandler handles unread message tracking endpoints.
type ReadStateHandler struct {
	readStateService services.ReadStateService
}

func NewReadStateHandler(readStateService services.ReadStateService) *ReadStateHandler {
	return &ReadStateHandler{readStateService: readStateService}
}

type markReadRequest struct {
	MessageID string `json:"message_id"`
}

// MarkRead marks a channel as read up to a specific message.
// POST /api/servers/{serverId}/channels/{id}/read
func (h *ReadStateHandler) MarkRead(w http.ResponseWriter, r *http.Request) {
	channelID := r.PathValue("id")

	user, ok := r.Context().Value(UserContextKey).(*models.User)
	if !ok {
		pkg.ErrorWithMessage(w, http.StatusUnauthorized, "user not found in context")
		return
	}

	var req markReadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		pkg.ErrorWithMessage(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if err := h.readStateService.MarkRead(r.Context(), user.ID, channelID, req.MessageID); err != nil {
		pkg.Error(w, err)
		return
	}

	pkg.JSON(w, http.StatusOK, map[string]string{"message": "marked as read"})
}

// MarkAllRead marks all channels in the server as read.
// POST /api/servers/{serverId}/channels/read-all
func (h *ReadStateHandler) MarkAllRead(w http.ResponseWriter, r *http.Request) {
	user, ok := r.Context().Value(UserContextKey).(*models.User)
	if !ok {
		pkg.ErrorWithMessage(w, http.StatusUnauthorized, "user not found in context")
		return
	}

	serverID, ok := r.Context().Value(ServerIDContextKey).(string)
	if !ok || serverID == "" {
		pkg.ErrorWithMessage(w, http.StatusBadRequest, "server context required")
		return
	}

	if err := h.readStateService.MarkAllRead(r.Context(), user.ID, serverID); err != nil {
		pkg.Error(w, err)
		return
	}

	pkg.JSON(w, http.StatusOK, map[string]string{"message": "all channels marked as read"})
}

type markMentionSeenRequest struct {
	MentionMessageID string `json:"mention_message_id"`
}

// MarkMentionSeen advances the mention-seen watermark for a channel.
// POST /api/servers/{serverId}/channels/{id}/read/mentions
func (h *ReadStateHandler) MarkMentionSeen(w http.ResponseWriter, r *http.Request) {
	channelID := r.PathValue("id")

	user, ok := r.Context().Value(UserContextKey).(*models.User)
	if !ok {
		pkg.ErrorWithMessage(w, http.StatusUnauthorized, "user not found in context")
		return
	}

	var req markMentionSeenRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		pkg.ErrorWithMessage(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if err := h.readStateService.MarkMentionSeen(r.Context(), user.ID, channelID, req.MentionMessageID); err != nil {
		pkg.Error(w, err)
		return
	}

	pkg.JSON(w, http.StatusOK, map[string]string{"message": "mention seen"})
}

// GetUnreads returns unread message counts for all channels in the server.
// GET /api/servers/{serverId}/channels/unread
func (h *ReadStateHandler) GetUnreads(w http.ResponseWriter, r *http.Request) {
	user, ok := r.Context().Value(UserContextKey).(*models.User)
	if !ok {
		pkg.ErrorWithMessage(w, http.StatusUnauthorized, "user not found in context")
		return
	}

	serverID, ok := r.Context().Value(ServerIDContextKey).(string)
	if !ok || serverID == "" {
		pkg.ErrorWithMessage(w, http.StatusBadRequest, "server context required")
		return
	}

	unreads, err := h.readStateService.GetUnreadCounts(r.Context(), user.ID, serverID)
	if err != nil {
		pkg.Error(w, err)
		return
	}

	pkg.JSON(w, http.StatusOK, unreads)
}

File: client/src/api/readState.ts
/**
 * Read State API — unread message tracking, server-scoped.
 */

import { apiClient } from "./client";

export type UnreadInfo = {
  channel_id: string;
  unread_count: number;
  last_mention_seen_at: string | null;
  last_mention_seen_message_id: string | null;
};

/** Marks a channel as read up to the given message ID. */
export function markRead(serverId: string, channelId: string, messageId: string) {
  return apiClient<{ message: string }>(`/servers/${serverId}/channels/${channelId}/read`, {
    method: "POST",
    body: { message_id: messageId },
  });
}

/** Returns unread counts for all channels in a server (channels with unread > 0 OR mention watermark set). */
export function getUnreadCounts(serverId: string) {
  return apiClient<UnreadInfo[]>(`/servers/${serverId}/channels/unread`);
}

/** Marks all channels in a server as read. */
export function markAllRead(serverId: string) {
  return apiClient<{ message: string }>(`/servers/${serverId}/channels/read-all`, {
    method: "POST",
  });
}

/** Advances the per-channel mention-seen watermark to the given mention message. */
export function markMentionSeen(serverId: string, channelId: string, mentionMessageId: string) {
  return apiClient<{ message: string }>(`/servers/${serverId}/channels/${channelId}/read/mentions`, {
    method: "POST",
    body: { mention_message_id: mentionMessageId },
  });
}
